sw/snRuntime/src/perf_cnt.h

```diff
@@ -2,100 +2,55 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
-#pragma once
+#define SNRT_NUM_PERF_CNTS SNITCH_CLUSTER_PERIPHERAL_PARAM_NUM_PERF_COUNTERS
 
-/// Different perf counters
-// Must match with `snitch_cluster_peripheral`
-enum snrt_perf_cnt {
-    SNRT_PERF_CNT0,
-    SNRT_PERF_CNT1,
-    SNRT_PERF_CNT2,
-    SNRT_PERF_CNT3,
-    SNRT_PERF_CNT4,
-    SNRT_PERF_CNT5,
-    SNRT_PERF_CNT6,
-    SNRT_PERF_CNT7,
-    SNRT_PERF_CNT8,
-    SNRT_PERF_CNT9,
-    SNRT_PERF_CNT10,
-    SNRT_PERF_CNT11,
-    SNRT_PERF_CNT12,
-    SNRT_PERF_CNT13,
-    SNRT_PERF_CNT14,
-    SNRT_PERF_CNT15,
-    SNRT_PERF_N_CNT,
-};
-
-/// Different types of performance counters
-enum snrt_perf_cnt_type {
-    SNRT_PERF_CNT_CYCLES,
-    SNRT_PERF_CNT_TCDM_ACCESSED,
-    SNRT_PERF_CNT_TCDM_CONGESTED,
-    SNRT_PERF_CNT_ISSUE_FPU,
-    SNRT_PERF_CNT_ISSUE_FPU_SEQ,
-    SNRT_PERF_CNT_ISSUE_CORE_TO_FPU,
-    SNRT_PERF_CNT_RETIRED_INSTR,
-    SNRT_PERF_CNT_RETIRED_LOAD,
-    SNRT_PERF_CNT_RETIRED_I,
-    SNRT_PERF_CNT_RETIRED_ACC,
-    SNRT_PERF_CNT_DMA_AW_STALL,
-    SNRT_PERF_CNT_DMA_AR_STALL,
-    SNRT_PERF_CNT_DMA_R_STALL,
-    SNRT_PERF_CNT_DMA_W_STALL,
-    SNRT_PERF_CNT_DMA_BUF_W_STALL,
-    SNRT_PERF_CNT_DMA_BUF_R_STALL,
-    SNRT_PERF_CNT_DMA_AW_DONE,
-    SNRT_PERF_CNT_DMA_AW_BW,
-    SNRT_PERF_CNT_DMA_AR_DONE,
-    SNRT_PERF_CNT_DMA_AR_BW,
-    SNRT_PERF_CNT_DMA_R_DONE,
-    SNRT_PERF_CNT_DMA_R_BW,
-    SNRT_PERF_CNT_DMA_W_DONE,
-    SNRT_PERF_CNT_DMA_W_BW,
-    SNRT_PERF_CNT_DMA_B_DONE,
-    SNRT_PERF_CNT_DMA_BUSY,
-    SNRT_PERF_CNT_ICACHE_MISS,
-    SNRT_PERF_CNT_ICACHE_HIT,
-    SNRT_PERF_CNT_ICACHE_PREFETCH,
-    SNRT_PERF_CNT_ICACHE_DOUBLE_HIT,
-    SNRT_PERF_CNT_ICACHE_STALL,
-};
+typedef struct {
+    uint16_t hart;
+    uint16_t metric;
+} perf_cnt_cfg_t;
 
 typedef union {
     uint32_t value __attribute__((aligned(8)));
+    perf_cnt_cfg_t cfg __attribute__((aligned(8)));
 } perf_reg32_t;
 
 typedef struct {
-    volatile perf_reg32_t enable[SNRT_PERF_N_CNT];
-    volatile perf_reg32_t hart_select[SNRT_PERF_N_CNT];
-    volatile perf_reg32_t perf_counter[SNRT_PERF_N_CNT];
+    volatile perf_reg32_t enable[SNRT_NUM_PERF_CNTS];
+    volatile perf_reg32_t select[SNRT_NUM_PERF_CNTS];
+    volatile perf_reg32_t perf_counter[SNRT_NUM_PERF_CNTS];
 } perf_regs_t;
```

Comment:
Indeed it made sense (and I think it's better) to have a union with a struct internally. But this is only needed on the right-hand side of the assignment in `snrt_cfg_perf_counter()`, where you now use the `<<` and `|` ops instead.

If we really need the `perf_reg32_t` type, can't we just use a `typedef` instead of a union? Accessing `.value` everywhere seems unnecessary to me. But I also still don't understand why we need this custom type at all rather than just a common `uint32_t`.

Reply:
The custom `perf_reg32_t` is useful here because all the configuration registers are 8 byte aligned, eventhough only the lower 32 bits are needed to configure the performance counter (the actual performance counter are 48-bit). With the `perf_reg32_t` arrays you can directly index with the performance counter index passed to the functions, which is kind of neat imo. I also tried with a standard `typedef` but then it only aligns the start of the arrays by 8-byte but not all the elements.

Reply:
Clear thanks :)